[PF3] Select Component

glekner opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->

**What**:
WIP for a Select Component with a Search Input.
I took the design from [here](https://www.patternfly.org/pattern-library/widgets/#bootstrap-select) even though there is no search bar.
<img width="265" alt="Screenshot 2019-03-27 17 35 02" src="https://user-images.githubusercontent.com/24938324/55089796-ab511b00-50b6-11e9-898a-7d9c396b1bf6.png">

Right now the Component accept for options an object e.g. 
```
{
   '1': 'one',
   '2': 'two',
    ...
}
```
<!-- Are there any upstream issues or separate issues you need to reference? -->

**TODO**:
- tests
- css variables
- more features?
<!-- feel free to add additional comments -->

patternfly-build commented:
PatternFly-React preview: https://1650-pr-patternfly-react-patternfly.surge.sh

glekner commented:
- added a disabled prop to an option, so now the options object looks like this
`{ '1': { label: 'one', disabled: false } ... } `

Hyperkid123 commented:
Can we change the format to an Array instead of Object? 

Something like:
```javascript
[{
  label: 'Foo',
  value: 1,
  ...
}, ...]
```

I mean if you want to render anything from Object in React you have to convert it into Array anyway.

glekner commented:
> Can we change the format to an Array instead of Object?
> 
> Something like:
> 
> ```js
> [{
>   label: 'Foo',
>   value: 1,
>   ...
> }, ...]
> ```
> 
> I mean if you want to render anything from Object in React you have to convert it into Array anyway.
> If we were to use this in existing applications we would have to transform all of our select options into Objects.

@Hyperkid123 
I Agree, but Many apps use JQuery Select that accepts an Object so I thought we might want to support that. I'm fine with switching to array

Hyperkid123 commented:
I don't think we want to model our components on jQuery at all.

glekner commented:
- Select now accepts an Array
- Fixed css bugs and overflows

@Hyperkid123 @jeff-phillips-18 @serenamarie125 can you please try the component?
we need this component before the freeze. Thanks!

Hyperkid123 commented:
Do you plan adding a multi select version of this?

glekner commented:
> Do you plan adding a multi select version of this?
In the future if its ok.

> How do you plan to handle clearing selected value? Adding an option with empty value or adding some icon which would clear the value?
Current Select designs doesn't have this feature - clearing the value. Would be able to add it in a future PR.

> Also i think that the select should expand when the selected option label is longer than the value container. It should extend to 100% width of parent element. @terezanovotna can tell you more about the actual PF pattern.
K, working on it.

glekner commented:
- Changed width to fill 100% of container
- Added `onKeyDown` and `cursor` props, the Storybook now responds to `up` `down` and `enter` keys upon focusing the search input

terezanovotna commented:
Hey @glekner happy to review ux for this PR, when you have a storybook, feel free to ping me!

Awesome you are adding search into the multi select. 
I suggest having the same design and behavior as [this component](https://www.patternfly.org/pattern-library/widgets/#bootstrap-select) you suggested. 

When more than 4 items selected - we can display `4 selected` instead of naming them all as there is not enough space for it.

glekner commented:
Thanks @terezanovotna ! the storybook the PR created is a bit outdated but the design is about the same, [try it](https://1650-pr-patternfly-react-patternfly.surge.sh/patternfly-3/index.html?selectedKind=patternfly-react%2FWidgets%2FAsyncSelect&selectedStory=AsyncSelect&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

@Hyperkid123 anyway to update the PR's storybook?

Hyperkid123 commented:
> @Hyperkid123 anyway to update the PR's storybook?

It should update once all checks are fine (no conflicts tests etc.)

glekner commented:
> > @Hyperkid123 anyway to update the PR's storybook?
> 
> It should update once all checks are fine (no conflicts tests etc.)

Thanks, any idea why travis isn't running?

Hyperkid123 commented:
Try close/open the PR.

Travis is acting up lately (on every project)

Hyperkid123 commented:
Hmm did not help i will ask someone to reset travis.

glekner commented:
added `disabled` prop for the Select, and added snapshot file.
travis still doesn't seem to run after rebase..

Hyperkid123 commented:
I have no clue what is going on wit travis. We restated it manually yesterday. Maybe we could try deleting cache on travis.

terezanovotna commented:
the styling of filter (blue border) should reflect [pf style](https://www.patternfly.org/pattern-library/forms-and-controls/find/#code)

As @Hyperkid123 mentioned it would be great to have multi select.

dtaylor113 commented:
Hi @Hyperkid123, looks great!   Components which do not have PatternFly designs/patterns should go into `patternfly-react-extensions` package.  -thanks

Hyperkid123 commented:
@dtaylor113  i actually did not write any code just trying to help @glekner

glekner commented:
@dtaylor113 
- Added `isSearchable` prop to control showing the search input
- Moved to `patternfly-react-extensions`

Can we merge? :)

dtaylor113 commented:
> Can we merge? :)

We need another approving review

glekner commented on packages/patternfly-3/patternfly-react-extensions/src/components/Select/Select.js:
```diff
@@ -0,0 +1,181 @@
+    results.length === 0 ? (
+      <ListGroupItem id="select-empty" key="empty" className="select-empty-list">
+        <span id="empty-text">{emptyText}</span>
+      </ListGroupItem>
```
This is not a separate component, only a function for better readability, pretty common in this library.

Laviro replied:
I meant seperating it, for example: 
- EmptyState
- ListItems
And other parts of the code that you could break into smaller components or helpers file.

glekner commented:
@dtaylor113 @jeff-phillips-18 can you ack? Thanks

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+@select-max-height: 205px;
+@select-min-width: 220px;
+
+.select-container {
```
Should have a `-pf` suffix

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+  overflow: hidden;
+  width: 100%;
+  min-width: @select-min-width;
+  border: 1px solid #ccc;
```
Can we use a color from the palette here?

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+      position: relative;
+
+      input {
+        width: 100% !important;
```
this is a `form-control` which already has width: 100%. Why this and why the `!important`?

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+
+        &:focus {
+          background-color: @color-pf-blue-400;
+          color: #fff;
```
color-pf-white

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/src/components/Select/Select.js:
```diff
@@ -0,0 +1,176 @@
+
+Select.propTypes = {
+  /** cursor position for keyboard navigation */
+  cursor: PropTypes.number,
```
why are we controlling keyboard navigation via the arrow keys rather than allowing the native control (tab)?

glekner replied:
who said it has to be the arrow keys? its arrow keys just for the storybook example.

jeff-phillips-18 replied:
ok, so what is the point in general?

glekner replied:
the arrow keys is a good example but it doesn't scroll if the cursor is going below the visible list. Maybe its best to remove this functionality for now and add it in a future PR, we want this hopefully to be merged today. Thoughts?

jeff-phillips-18 replied:
I'd be ok with removing it and adding if/when its deemed necessary

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+
+    .select-toggle-placeholder {
+      font-style: italic;
+      color: #9c9c9c;
```
palette color

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+        &#select-empty {
+          user-select: none;
+          font-style: italic;
+          color: #9c9c9c;
```
palette color

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/less/select.less:
```diff
@@ -0,0 +1,141 @@
+          text-align: center;
+          user-select: none;
+          font-style: italic;
+          color: #9c9c9c;
```
palette color

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/src/components/Select/Select.js:
```diff
@@ -0,0 +1,176 @@
+  selectedItem: {
+    id: '10',
+    name: 'ten'
+  }
```
could cause issues for applications not supplying a selectedItem, if they happen to have a matching item it will be inadvertently selected.

glekner commented:
implemented @jeff-phillips-18 suggestions and removed `cursor` functionality for now.

@dtaylor113 @jeff-phillips-18 ack?

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react-extensions/src/components/Select/StatefulWrapperSelect.js:
```diff
@@ -0,0 +1,101 @@
+class StatefulWrapperSelect extends Component {
+  state = {
+    open: false,
+    cursor: 0,
```
Should probably remove the cursor code within this class to avoid confusions.

glekner commented:
@jeff-phillips-18 removed `cursor` from stateful wrapper

jeff-phillips-18 commented:
@glekner Could you please reword your commit?

jeff-phillips-18 commented:
Looks like you need to update your snapshots:

```
Summary of all failing tests
 FAIL  packages/patternfly-3/patternfly-react-extensions/src/components/Select/Select.test.js
  ● Select renders properly

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot "Select renders properly 1".

    - Snapshot
    + Received

    @@ -1,7 +1,6 @@
      <Select
    -   cursor={-1}
        disabled={false}
        emptyText="No Results"
        focusOnMount={true}
        isLoading={false}
        isSearchable={true}

      16 |   ).getElement();
      17 | 
    > 18 |   expect(component).toMatchSnapshot();
         |                     ^
      19 | });
      20 | 
      21 | test('Select functionality', () => {

      at Object.toMatchSnapshot (packages/patternfly-3/patternfly-react-extensions/src/components/Select/Select.test.js:18:21)
```

glekner commented:
fixed commit and snapshot @jeff-phillips-18 @dtaylor113
